18F/identity-idp
# LG-10286: Allow AAL2-restricted sign in to choose another option

## 🎫 Ticket

[LG-10286](https://cm-jira.usa.gov/browse/LG-10286)

## 🛠 Summary of changes

Updates AAL2 strict authentication requests to use a more standardized user flow, allowing a user to use all available methods supported for the request.

Previously, it would not be possible to use Face or Touch Unlock to sign in to a partner requesting phishing-resistant MFA.

With these changes, the user will always be given the option to "Choose another authentication method" when prompted for their MFA. If MFA options are limited due to request parameters of the partner, a warning will be shown on the MFA selection screen explaining why the options are limited.

**Draft:** Merges to #8834

## 📜 Testing Plan

1. Go to http://localhost:3000
2. Create an account with PIV/CAC as your MFA
3. From account dashboard, click "Add Face or Touch Unlock" and complete enrollment
4. From account dashboard, click "Security key' and complete enrollment
5. From account dashboard, click "Forget all browsers" and complete this flow
6. Sign out
7. With [example sample app](https://github.com/18f/identity-oidc-sinatra) running in a separate process, go to http://localhost:9292
8. Under Options, select "Phishing-resistant AAL2" under "Authentication Assurance Level (AAL)"
9. Click "Sign in"
10. Sign in to the account you created
11. Confirm that you are able to use any of your MFA methods to sign in

## 👀 Screenshots

**Phishing Resistant Required:**

Screen|Before|After
---|---|---
MFA selection|N/A|![mfa-select-after](https://github.com/18F/identity-idp/assets/1779930/75876a4a-cc68-4c26-b18b-ae81a4fcc4ff)
PIV authentication|![piv-before](https://github.com/18F/identity-idp/assets/1779930/6967bdce-5fb1-4502-a143-e052956ca230)|![piv-after](https://github.com/18F/identity-idp/assets/1779930/4b08e8b6-92f4-4365-84e1-40d271e67ea5)
Security Key authentication|![security-key-before](https://github.com/18F/identity-idp/assets/1779930/0a6963e1-a347-4b28-8c7f-ae8104a60ff4)|![security-key-after](https://github.com/18F/identity-idp/assets/1779930/a43a98e5-dbb6-4862-bb37-719f7b96807c)
Face/Touch authentication|N/A|![ft-after](https://github.com/18F/identity-idp/assets/1779930/c5e5b314-2b37-4ed4-bdf6-c4de49d34433)

**PIV/CAC Only:**

Screen|Before|After
---|---|---
PIV authentication|![piv-only-before](https://github.com/18F/identity-idp/assets/1779930/71ffe7cb-3629-4a5b-ba9a-b0c05a30e1ee)|![piv-only-after](https://github.com/18F/identity-idp/assets/1779930/9b720af7-035e-4e16-a770-e5790f0fd062)
MFA selection|N/A|![piv-only-mfa-select-after](https://github.com/18F/identity-idp/assets/1779930/20eed338-140b-4264-b97b-57b0b75f2ea0)
